client/DESIGN.md

```diff
@@ -0,0 +1,380 @@
+
+We use the following HTTP directives (under the HTTP header Cache-Control):
+
+* **max-age**: The max number of seconds between our given read\_timestamp and the read\_timestamp of the entry, i.e., (input read\_timestamp) - (entry read\_timestamp) <= max-age.  In other words, we are willing to read a value that could be stale by at most max-age seconds. 
+
+* **no-cache**: Forces cache to send a request to the DB server to retrieve the requested value.  The result cannot just be returned from cache.
+
+An example of usage is below:
+
+    GET /[table_id_string]/[key_string] HTTP/1.1
+    Cache-Control: max-age=3, no-cache
+
+In addition, we use the following custom HTTP header: 
+
+* **Read-TxClock**: Contains the user's specified read\_timestamp.  The value returned by the DB server will be the most recent value older than the specified read\_timestamp value.  Note that Read-TxClock is an long (64 bits).
+
+Read-TxClock is specified in microseconds to clarify that it should **not** be *manipulated* as a date.  While applications may display Read-TxClock to users in Date format for clarity, the application must internally maintain Read-TxClock as a long (64 bits) to avoid rounding to seconds. 
+
+Additionally, note that max-age and no-cache are existing HTTP Caching headers specified in [RFC7234](http://tools.ietf.org/html/rfc7234#section-5.2.1), while the read\_timestamp is sent through the Read-TxClock header, a custom HTTP header.
+
+Below are the expected forms of requests/responses between the cache and TreodeDB server:
+
+### Requests
+
+We give examples of HTTP requests below.  For additional explanation, please see the source of these examples [OMVCC and Caching](https://forum.treode.com/t/omvcc-and-http-caching/62).
+
+#### Read Requests  
+
+The cache message to the DB includes the table id, the desired key, and headers restricting the timestamps of the value.  (We describe in the Use of HTTP Headers section how each header restricts the value we read.)
+
+HTTP read requests should follow this format:
+
+    GET /[table_id_string]/[key_string] HTTP/1.1
+    Cache-Control: max-age=[max_age_in_seconds], no-cache
+    Read-TxClock: [read_timestamp_in_microseconds_long_64_bits]
+
+Here is an example of an HTTP read request:
+
+    GET /movie/star-wars HTTP/1.1
+    Cache-Control: max-age=300, no-cache
+    Read-TxClock: 1421236153024853
+
+The user is requesting the value of key "star-wars" in the table "movie".  Since this key's value may have changed over time, the user gives a time range for the value they want.  
+
+Specifically, the user wants a value that was written to the DB by time 1421236153024853 microseconds (where time 0 is some arbitrary starting point in the past).  However, the user does not want a value that was written to the DB more than 300 seconds before time 1421236153024853.  Thus, the time range is 
```

Comment:
The `max-age` directive is for intermediate caches, the DB ignores it.

> However, the user does not want a value that was written to the DB more than 300 seconds before time 1421236153024853.

The user wants a value written on or before time 1421236153024853. They are willing to accept the cached value from an intermediate HTTP proxying cache if that cache read that value more recently than the last 300 seconds.